rllib/core/rl_module/marl_module.py

```diff
@@ -0,0 +1,295 @@
+from typing import Iterator, Mapping, Any, Union, Dict, Set
+import pprint
+
+
+from ray.rllib.utils.annotations import (
+    OverrideToImplementCustomLogic,
+    override,
+)
+
+from ray.rllib.models.specs.specs_dict import ModelSpec
+from ray.rllib.policy.sample_batch import MultiAgentBatch
+from ray.rllib.core.rl_module import RLModule
+
+ModuleID = str
+
+
+class MultiAgentRLModule(RLModule):
+    """Base class for multi-agent RLModules.
+
+    This class holds a mapping from module_ids to RLModules. It provides a convenient
+    way of accessing each individual module as well as all of with only one call.
+
+    The default implementation assumes the data communicated as input and output of
+    RLModule APIs are multi-agent batches. Each RLModule simply takes its own input and
+    produces its own output. It also assumes that the RLModules do not share any
+    parameters or communication with one another. The behavior of modules with such
+    advanced communication would be undefined. To share parameters or communication
+    between RLModules, you should implement your own MultiAgentRLModule.
+
+    Input config keys:
+        `modules`: Mapping from module_id to RLModule config.
+        `trainable_modules`: Set of module_ids that are trainable. If not specified,
+            all modules are trainable.
+
+    RLModule config values can be of the following forms:
+        1. The RLModule instance itself.
+        2. A dict with the following keys:
+            `module_class`: The RLModule class Type or full path separate by `.`.
+                (e.g. `ray.rllib.algoirhms.dqn.torch.DQNTorchRLModule`).
+            `module_config`: The config object for the RLModule.
+        3. A tuple of (RLModule class Type, config object). The RLModule class Type can
+            be the type itself or the full path separate by `.`.
+
+
+    Args:
+        config: The config dict for the multi-agent RLModule (see above).
+    """
+
+    def __init__(self, config: Mapping[str, Any], **kwargs) -> None:
+        super().__init__(config, **kwargs)
+
+        self._rl_modules: Mapping[ModuleID, RLModule] = self._make_modules()
+        self._trainable_rl_modules = self.config.get(
+            "trainable_modules", set(self._rl_modules.keys())
+        )
```

Comment:
today, this is a function that returns a mapping from id to bools right?
there may be a good reasons it works that way.
will this have limitations?
also, type annotation for _trainable_rl_modules?

Reply:
+1
This should be 100% analogous to what we currently do for trainable Policies.

* None: All modules are trainable, even later dynamically added ones
* some container (list, tuple, set): These module IDs are trainable.
* callable[ModuleID, batch=None] -> bool: callable that takes a Module ID and (optionally) a MABatch and returns whether that ID is trainable under the condition of the given MA batch.

Reply:
made a few changes here:
1) internal rl_module and trainable_rl_modules are private variables with `__`.
2) Added a type annotation for what `__trainable_rl_modules` are: `Set[ModuleID]`
3) If None --> all modules are trainable. the modules that are added later are not trainable only if `is_trainable=False` is passed in when adding module with `add_module`
4) right now I didn't think about the case where the config passes in a callable that controls the logic of what is trainable and what's not. I added a todo to add this later. But a question regarding this @sven1977 What is the use-cases of passing MABatch to is_trainable of a module?

Reply:
👍 
this is a chance to do things right. it's good to question some of the design decisions that seem irrelevant now.

Reply:
4. The use case is that you might want to train certain module **only** if they have collected samples while playing against a particular opponent module. So you need to peek into the batch/collected data to figure this out.

Reply:
Wait until I add this functionality. I think it's kinda important to capture now as it implies changes to the public API